appengine/js/blocks.js

```diff
@@ -0,0 +1,452 @@
+      } else if (thisBlock.elseifCount_ && !thisBlock.elseCount_) {
+        return Blockly.Msg.CONTROLS_IF_TOOLTIP_3;
+      } else if (thisBlock.elseifCount_ && thisBlock.elseCount_) {
+        return Blockly.Msg.CONTROLS_IF_TOOLTIP_4;
+      }
+      return '';
+    });
+    this.elseifCount_ = 0;
+    this.elseCount_ = 0;
+  },
+  /**
+   * Modify this block to have the correct number of inputs.
+   * @private
+   * @this Blockly.Block
+   */
+  updateShape_: function() {
+    // Delete everything.
+    if (this.getInput('ELSE')) {
+      this.removeInput('ELSEMSG');
+      this.removeInput('ELSE');
+    }
+    var i = 1;
+    while (this.getInput('IF' + i)) {
+      this.removeInput('IF' + i);
+      this.removeInput('TAIL' + i);
+      this.removeInput('DO' + i);
+      i++;
+    }
+    // Rebuild block.
+    for (var i = 1; i <= this.elseifCount_; i++) {
+      this.appendValueInput('IF' + i)
+          .setCheck('Boolean')
+          .appendField('} else if (');
+      this.appendDummyInput('TAIL' + i)
+          .appendField(') {');
+      this.appendStatementInput('DO' + i);
+    }
+    if (this.elseCount_) {
+      this.appendDummyInput('ELSEMSG')
+          .appendField('} else {');
+      this.appendStatementInput('ELSE');
+    }
+    // Move final '}' to the end.
+    this.moveInputBefore('TAIL', null);
+  },
+  mutationToDom: Blockly.Blocks['controls_if'].mutationToDom,
+  domToMutation: Blockly.Blocks['controls_if'].domToMutation,
+  decompose: Blockly.Blocks['controls_if'].decompose,
+  compose: Blockly.Blocks['controls_if'].compose,
+  saveConnections: Blockly.Blocks['controls_if'].saveConnections
+};
+
+/**
+ * Comparison operator.
+ * @this Blockly.Block
+ */
+Blockly.Blocks['logic_compare'].init = function() {
+  var OPERATORS = [
+    ['==', 'EQ'],
+    ['!=', 'NEQ'],
+    ['<', 'LT'],
+    ['<=', 'LTE'],
+    ['>', 'GT'],
+    ['>=', 'GTE']
+  ];
+  this.setHelpUrl(Blockly.Msg.LOGIC_COMPARE_HELPURL);
+  this.setColour(Blockly.Blocks.logic.HUE);
+  this.setOutput(true, 'Boolean');
+  this.appendValueInput('A');
+  this.appendValueInput('B')
+      .appendField(new Blockly.FieldDropdown(OPERATORS), 'OP');
+  this.setInputsInline(true);
+  // Assign 'this' to a variable for use in the tooltip closure below.
+  var thisBlock = this;
+  this.setTooltip(function() {
+    var op = thisBlock.getFieldValue('OP');
+    var TOOLTIPS = {
+      EQ: Blockly.Msg.LOGIC_COMPARE_TOOLTIP_EQ,
+      NEQ: Blockly.Msg.LOGIC_COMPARE_TOOLTIP_NEQ,
+      LT: Blockly.Msg.LOGIC_COMPARE_TOOLTIP_LT,
+      LTE: Blockly.Msg.LOGIC_COMPARE_TOOLTIP_LTE,
+      GT: Blockly.Msg.LOGIC_COMPARE_TOOLTIP_GT,
+      GTE: Blockly.Msg.LOGIC_COMPARE_TOOLTIP_GTE
+    };
+    return TOOLTIPS[op];
+  });
+  this.prevBlocks_ = [null, null];
+};
+
+Blockly.Msg.LOGIC_OPERATION_AND = '&&';
+Blockly.Msg.LOGIC_OPERATION_OR = '||';
+
+Blockly.Msg.LOGIC_BOOLEAN_TRUE = 'true';
+Blockly.Msg.LOGIC_BOOLEAN_FALSE = 'false';
+
+Blockly.Blocks['controls_whileUntil'] = {
```

Comment:
The controls_whileUntil block (along with controls_if, math_arithmetic, and math_change) could all be simplified by setting init directly, as is the case with the rest of the blocks (e.g. variables_get/set).  The controls_if block would also need to set updateShape_.